From 8384a63d89be057c8d0d9565e26b8c923eb7aad8 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 9 Dec 2024 17:48:19 +0100
Subject: [PATCH 1452/1695] drivers: stm32_iwdg: remove useless device list

STM32 watchdog driver does not manage several instances of IWDG hence
remove the useless code. To simplify code, remove stm32_iwdg_register()
local function.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 60f8d871431695bea180142089618b0277b60c1a drivers: stm32_iwdg: remove useless device list]
Change-Id: I3c6a30e6835ec75aef51f5780dc7ad290ca1971c
Link: https://github.com/OP-TEE/optee_os/commit/a096e2d9c573770a6c765d89bf1b30f276278610
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/434835
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/stm32_iwdg.c | 47 +++++++++++----------------------------
 1 file changed, 13 insertions(+), 34 deletions(-)

diff --git a/core/drivers/stm32_iwdg.c b/core/drivers/stm32_iwdg.c
index f6a3adca8..ebed7f560 100644
--- a/core/drivers/stm32_iwdg.c
+++ b/core/drivers/stm32_iwdg.c
@@ -96,7 +96,6 @@
  * @hw_version - Watchdog HW version
  * @last_refresh - Time of last watchdog refresh
  * @wdt_chip - Wathcdog chip instance
- * @link - Link in registered watchdog instance list
  * @max_hw_timeout - Maximum hardware timeout
  */
 struct stm32_iwdg_device {
@@ -114,13 +113,9 @@ struct stm32_iwdg_device {
 	unsigned int hw_version;
 	TEE_Time last_refresh;
 	struct wdt_chip wdt_chip;
-	SLIST_ENTRY(stm32_iwdg_device) link;
 	unsigned long max_hw_timeout;
 };
 
-static SLIST_HEAD(iwdg_dev_list_head, stm32_iwdg_device) iwdg_dev_list =
-	SLIST_HEAD_INITIALIZER(iwdg_dev_list_head);
-
 static uint32_t sr_ewif_mask(struct stm32_iwdg_device *iwdg)
 {
 	if (iwdg->hw_version >= IWDG_ICR_MIN_VER)
@@ -540,41 +535,23 @@ static TEE_Result stm32_iwdg_setup(struct stm32_iwdg_device *iwdg,
 	return TEE_SUCCESS;
 }
 
-static TEE_Result
-stm32_iwdg_pm(enum pm_op op, unsigned int pm_hint __unused,
-	      const struct pm_callback_handle *pm_handle __unused)
+static TEE_Result stm32_iwdg_pm(enum pm_op op, unsigned int pm_hint __unused,
+				const struct pm_callback_handle *pm_handle)
 {
-	struct stm32_iwdg_device *iwdg = NULL;
+	struct stm32_iwdg_device *iwdg = PM_CALLBACK_GET_HANDLE(pm_handle);
 
-	SLIST_FOREACH(iwdg, &iwdg_dev_list, link) {
-		if (op == PM_OP_RESUME) {
-			clk_enable(iwdg->clk_lsi);
-			clk_enable(iwdg->clk_pclk);
-		} else {
-			clk_disable(iwdg->clk_lsi);
-			clk_disable(iwdg->clk_pclk);
-		}
+	if (op == PM_OP_RESUME) {
+		clk_enable(iwdg->clk_lsi);
+		clk_enable(iwdg->clk_pclk);
+	} else {
+		clk_disable(iwdg->clk_lsi);
+		clk_disable(iwdg->clk_pclk);
 	}
 
 	return TEE_SUCCESS;
 }
 DECLARE_KEEP_PAGER_PM(stm32_iwdg_pm);
 
-static TEE_Result stm32_iwdg_register(struct stm32_iwdg_device *iwdg)
-{
-	TEE_Result res = TEE_ERROR_GENERIC;
-
-	iwdg->wdt_chip.ops = &stm32_iwdg_ops;
-
-	res = watchdog_register(&iwdg->wdt_chip);
-	if (res)
-		return res;
-
-	SLIST_INSERT_HEAD(&iwdg_dev_list, iwdg, link);
-
-	return TEE_SUCCESS;
-}
-
 static TEE_Result stm32_iwdg_probe(const void *fdt, int node,
 				   const void *compat_data __unused)
 {
@@ -589,11 +566,13 @@ static TEE_Result stm32_iwdg_probe(const void *fdt, int node,
 	if (res)
 		goto err;
 
-	res = stm32_iwdg_register(iwdg);
+	iwdg->wdt_chip.ops = &stm32_iwdg_ops;
+
+	res = watchdog_register(&iwdg->wdt_chip);
 	if (res)
 		goto err;
 
-	register_pm_core_service_cb(stm32_iwdg_pm, NULL, "stm32-iwdg");
+	register_pm_core_service_cb(stm32_iwdg_pm, iwdg, "stm32-iwdg");
 
 	return TEE_SUCCESS;
 
-- 
2.47.2

